### ISSUE TITLE
Nested empty record type gets pruned

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#9773]
A record type defined inside a function gets pruned when it shouldn't be. The simplest reproducer, motivated by the upcoming version of  cg-sparse-partred.chpl. is:

```chpl
proc plusPR1(FEXPR) {}

proc conjGrad(A) {
    record Empty { proc init(){} }
    plusPR1(new Empty());
}
proc main {
  conjGrad(Locales);
}
```

BTW this should work without `proc init()` inside `Empty`. Currently it does not, saying `could not generate default initializer for type 'Empty', please define one` . So this issue is also under the "Compiler-generated Initializers" epic.

#### Test

    test/types/records/init/nested-empty-record.chpl

#### Developer notes

The AggregateType for `Empty` gets pruned during resolution. At that point its fields are set so:

    aggregateTag = AGGREGATE_RECORD,
    unmanagedClass = 0x0,
    typeConstructor = 0x0,
    defaultInitializer = 0x0,
    instantiatedFrom = 0x0,
    initializerStyle = DEFINES_INITIALIZER,
    initializerResolved = false,

As a result, do_isUnusedClass() returns 'true' and the AggregateType gets pruned.

**Comment by cassella** [comment]
The code above passes with `--verify`, with or without the `proc init() {}`.

The added future, `test/types/records/init/nested-empty-record.chpl`, was retired in #12034

**Comment by vasslitvinov** [comment]
@benharsh - this was resolved in #12034, right?

**Comment by benharsh** [comment]
I think so, I'll close the issue.
